### PR TITLE
[mlir][Vector] Add patterns for efficient unsigned i4 -> i8 conversion emulation

### PR DESCRIPTION
**Pull request by KoolJBlack**
This PR builds on https://github.com/llvm/llvm-project/pull/79494 with an additional path for efficient unsigned `i4 ->i8` type extension for 1D/2D operations. This will impact any i4 -> i8/i16/i32/i64 unsigned extensions as well as sitofp i4 -> f8/f16/f32/f64.

### CODE REVIEW

File: mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir [hunk 1]
```diff
@@ -324,6 +324,50 @@ func.func @i7_transpose(%a: vector<8x16xi7>) -> vector<16x8xi7> {
...
+// CHECK:           %[[VAL_1:.*]] = arith.constant dense<4> : vector<4xi8>
+// CHECK:           %[[VAL_2:.*]] = arith.constant dense<15> : vector<4xi8>
+// CHECK:           %[[VAL_3:.*]] = vector.bitcast %[[VAL_0]] : vector<8xi4> to vector<4xi8>
+// CHECK:           %[[VAL_4:.*]] = arith.andi %[[VAL_3]], %[[VAL_2]] : vector<4xi8>
+// CHECK:           %[[VAL_5:.*]] = arith.shrsi %[[VAL_3]], %[[VAL_1]] : vector<4xi8>
+// CHECK:           %[[VAL_6:.*]] = vector.interleave %[[VAL_4]], %[[VAL_5]] : vector<4xi8>
+  %0 = arith.extui %a : vector<8xi4> to vector<8xi8>
+  return %0 : vector<8xi8>
+}
+
```
Comment (dcaballe) [inline mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir:370]: you mentioned `sitofp` in the description, which I understand you meant `uitofp`... but I don't see the test and support in this PR?

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 2]
```diff
@@ -880,6 +880,38 @@ static Value rewriteI4ToI8SignedExt(PatternRewriter &rewriter, Location loc,
...
+                                             lowBitsMaskValues);
+  constexpr int8_t highBitsToShift = 4;
+  auto highShiftValues = rewriter.create<arith::ConstantOp>(
+      loc, DenseElementsAttr::get(i8VecType, highBitsToShift));
+  Value high = rewriter.create<arith::ShRSIOp>(loc, i8Vector, highShiftValues);
+
+  // 3. Interleave low and high i8 elements.
+  return rewriter.create<vector::InterleaveOp>(loc, low, high);
+}
+
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:914]: It looks like this function is identical to the signed one except for the expansion. Could we have a single function where the expansion is done based on an `isSigned` flag or enum or something like that?

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 3]
```diff
@@ -880,6 +880,38 @@ static Value rewriteI4ToI8SignedExt(PatternRewriter &rewriter, Location loc,
...
+  //  byte are place in one vector and the high i4 elements in another vector.
+  constexpr unsigned char lowBitsMask = 15; // Equivalent to [0000IIII] bit mask
+  auto lowBitsMaskValues = rewriter.create<arith::ConstantOp>(
+      loc, DenseElementsAttr::get(i8VecType, lowBitsMask));
+  Value low = rewriter.create<arith::AndIOp>(loc, i8Vector.getType(), i8Vector,
+                                             lowBitsMaskValues);
+  constexpr int8_t highBitsToShift = 4;
+  auto highShiftValues = rewriter.create<arith::ConstantOp>(
+      loc, DenseElementsAttr::get(i8VecType, highBitsToShift));
+  Value high = rewriter.create<arith::ShRSIOp>(loc, i8Vector, highShiftValues);
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:909]: This shift should be unsigned, right?

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 4]
```diff
@@ -1099,6 +1131,50 @@ struct RewriteAlignedSubByteIntSignedExt : OpRewritePattern<ConversionOpType> {
...
+      return failure();
+
+    // Check general alignment preconditions.
+    if (failed(alignedConversionPrecondition(rewriter, srcVecType, dstVecType,
+                                             conversionOp)))
+      return failure();
+
+    // Perform the rewrite.
+    Value subByteExt =
+        rewriteI4ToI8UnsignedExt(rewriter, conversionOp.getLoc(), srcValue);
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:1169]: If you can create a common function as mentioned above, you should be able to reuse the same pattern. You can try something like:
```
template <typename ConversionOpType, bool isSigned>
struct RewriteAlignedSubByteIntExt : OpRewritePattern<ConversionOpType> {

}
```

**Review by dcaballe — COMMENTED** [review]
LG! A few comments

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 5]
```diff
@@ -1099,6 +1131,50 @@ struct RewriteAlignedSubByteIntSignedExt : OpRewritePattern<ConversionOpType> {
...
+/// avoid leaving LLVM to scramble with peephole optimizations.
+///
+/// For example:
+///    arith.extui %in : vector<8xi4> to vector<8xi32>
+///      is rewritten as
+///        %0 = vector.bitcast %in : vector<8xi4> to vector<4xi8>
+///        %1 = arith.andi %0, 15 : vector<4xi8>
+///        %2 = arith.shrsi %0, 4 : vector<4xi8>
+///        %3 = vector.interleave %1, %2 : vector<4xi8>
+///        %4 = arith.extsi %3 : vector<8xi8> to vector<8xi32>
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:1145]: Formatting is off. Also, `arith.extsi` -> `arith.extui` ;-)

```suggestion
///        arith.extui %in : vector<8xi4> to vector<8xi32>
///    is rewritten as
///        %0 = vector.bitcast %in : vector<8xi4> to vector<4xi8>
///        %1 = arith.andi %0, 15 : vector<4xi8>
///        %2 = arith.shrsi %0, 4 : vector<4xi8>
///        %3 = vector.interleave %1, %2 : vector<4xi8>
///        %4 = arith.extsi %3 : vector<8xi8> to vector<8xi32>
```

File: mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir [hunk 6]
```diff
@@ -324,6 +324,50 @@ func.func @i7_transpose(%a: vector<8x16xi7>) -> vector<16x8xi7> {
...
+// CHECK-LABEL: func.func @aligned_extui(
+func.func @aligned_extui(%a: vector<8xi4>) -> vector<8xi32> {
+// CHECK-SAME:                             %[[VAL_0:.*]]: vector<8xi4>) -> vector<8xi32> {
+// CHECK:           %[[VAL_1:.*]] = arith.constant dense<4> : vector<4xi8>
+// CHECK:           %[[VAL_2:.*]] = arith.constant dense<15> : vector<4xi8>
+// CHECK:           %[[VAL_3:.*]] = vector.bitcast %[[VAL_0]] : vector<8xi4> to vector<4xi8>
+// CHECK:           %[[VAL_4:.*]] = arith.andi %[[VAL_3]], %[[VAL_2]] : vector<4xi8>
+// CHECK:           %[[VAL_5:.*]] = arith.shrsi %[[VAL_3]], %[[VAL_1]] : vector<4xi8>
+// CHECK:           %[[VAL_6:.*]] = vector.interleave %[[VAL_4]], %[[VAL_5]] : vector<4xi8>
+// CHECK:           %[[VAL_7:.*]] = arith.extui %[[VAL_6]] : vector<8xi8> to vector<8xi32>
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/vector-rewrite-narrow-types.mlir:336]: [nit] Could you use more descriptive LIT variable names? Same as other tests in this file?

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp (hunk 3 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:909]: +1

**Review by banach-space — COMMENTED** [review]
Nice!

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 7]
```diff
@@ -880,6 +880,38 @@ static Value rewriteI4ToI8SignedExt(PatternRewriter &rewriter, Location loc,
...
+
+  // 1. Generate a bitcast vector<Xxi4> -> vector<X/2xi8>.
+  SmallVector<int64_t> i8VecShape = llvm::to_vector(srcVecType.getShape());
+  constexpr int64_t i4Toi8BitwidthFactor = 2;
+  i8VecShape.back() = i8VecShape.back() / i4Toi8BitwidthFactor;
+  auto i8VecType = VectorType::get(i8VecShape, rewriter.getI8Type());
+  Value i8Vector = rewriter.create<vector::BitCastOp>(loc, i8VecType, srcValue);
+
+  // 2 Extend the i4 elements using shifts & masking. Low i4 elemens of each
+  //  byte are place in one vector and the high i4 elements in another vector.
```
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:900]: nit: spelling/grammar
```suggestion
  // 2. Extend the i4 elements using shifts & masking. Low i4 elements of each
  //  byte are placed in one vector and the high i4 elements in another vector.
```

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 8]
```diff
@@ -880,6 +880,38 @@ static Value rewriteI4ToI8SignedExt(PatternRewriter &rewriter, Location loc,
...
+  constexpr unsigned char lowBitsMask = 15; // Equivalent to [0000IIII] bit mask
```
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:901]: nit: `I` -> `1`?
```suggestion
  constexpr unsigned char lowBitsMask = 15; // Equivalent to [00001111] bit mask
```

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 9]
```diff
@@ -880,6 +880,38 @@ static Value rewriteI4ToI8SignedExt(PatternRewriter &rewriter, Location loc,
...
+  auto lowBitsMaskValues = rewriter.create<arith::ConstantOp>(
+      loc, DenseElementsAttr::get(i8VecType, lowBitsMask));
+  Value low = rewriter.create<arith::AndIOp>(loc, i8Vector.getType(), i8Vector,
```
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:904]:
```suggestion
  Value low = rewriter.create<arith::AndIOp>(loc, i8VecType, i8Vector,
```

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp (hunk 8 above)
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:901]: use stdint type for consistency?
```suggestion
  constexpr uint8_t lowBitsMask = 15; // Equivalent to [0000IIII] bit mask
```

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 10]
```diff
@@ -1099,6 +1131,50 @@ struct RewriteAlignedSubByteIntSignedExt : OpRewritePattern<ConversionOpType> {
...
+struct RewriteAlignedSubByteIntUnsignedExt
+    : OpRewritePattern<ConversionOpType> {
+  using OpRewritePattern<ConversionOpType>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ConversionOpType conversionOp,
+                                PatternRewriter &rewriter) const override {
+    // Verify the preconditions.
+    Value srcValue = conversionOp.getIn();
+    auto srcVecType = dyn_cast<VectorType>(srcValue.getType());
+    auto dstVecType = dyn_cast<VectorType>(conversionOp.getType());
```
Comment (c-rhodes) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:1157]: should this be checking the cast? If not I think it should just use `cast`

**Pull request merged**